session/nontransactional.go

```diff
@@ -0,0 +1,492 @@
+			}
+			if cmp != 0 {
+				jobs = append(jobs, job{jobID: jobCount, start: currentStart, end: currentEnd, jobSize: currentSize})
+				jobCount++
+				currentSize = 0
+			}
+		}
+
+		// a new batch
+		if currentSize == 0 {
+			currentStart = *newStart.Clone()
+		}
+
+		currentSize += chk.NumRows()
+		currentEndPointer := chk.GetRow(chk.NumRows()-1).GetDatum(0, &rs.Fields()[0].Column.FieldType)
+		currentEnd = *currentEndPointer.Clone()
+	}
+
+	return jobs, nil
+}
+
+func buildSelectSQL(stmt *ast.NonTransactionalDeleteStmt, se Session) (*ast.TableName, string, *table.Column, error) {
+	// only use the first table
+	// TODO: return error if there are multiple tables
+	if stmt.DeleteStmt.TableRefs == nil || stmt.DeleteStmt.TableRefs.TableRefs == nil {
+		return nil, "", nil, errors.New("table reference is nil")
+	}
+	if stmt.DeleteStmt.TableRefs.TableRefs.Right != nil {
+		return nil, "", nil, errors.New("Non-transactional delete doesn't support multiple tables")
+	}
+	tableSource, ok := stmt.DeleteStmt.TableRefs.TableRefs.Left.(*ast.TableSource)
+	if !ok {
+		return nil, "", nil, errors.New("Non-transactional delete, table source not found")
+	}
+	tableName, ok := tableSource.Source.(*ast.TableName)
+	if !ok {
+		return nil, "", nil, errors.New("Non-transactional delete, table name not found")
+	}
+
+	// the shard column must be indexed
+	indexed, shardColumn, err := checkShardColumnIndexed(stmt, se, tableName)
+	if err != nil {
+		return nil, "", nil, err
+	}
+	if !indexed {
+		return nil, "", nil, errors.Errorf("Non-transactional delete, shard column %s is not indexed", stmt.ShardColumn.Name.L)
+	}
+
+	var sb strings.Builder
+	if stmt.DeleteStmt.Where != nil {
+		err := stmt.DeleteStmt.Where.Restore(format.NewRestoreCtx(format.DefaultRestoreFlags|
+			format.RestoreNameBackQuotes|
+			format.RestoreSpacesAroundBinaryOperation|
+			format.RestoreBracketAroundBinaryOperation|
+			format.RestoreStringWithoutCharset, &sb))
+		if err != nil {
+			return nil, "", nil, errors.Trace(err)
+		}
+	} else {
+		sb.WriteString("true")
+	}
+	// assure NULL values are placed first
+	selectSQL := fmt.Sprintf("select `%s` from `%s`.`%s` where %s order by IF(ISNULL(`%s`),0,1),`%s`",
+		stmt.ShardColumn.Name.O, tableName.DBInfo.Name.O, tableName.Name.O, sb.String(), stmt.ShardColumn.Name.O, stmt.ShardColumn.Name.O)
+	return tableName, selectSQL, shardColumn, nil
+}
+
+func checkShardColumnIndexed(stmt *ast.NonTransactionalDeleteStmt, se Session, tableName *ast.TableName) (bool, *table.Column, error) {
+	shardColumnName := stmt.ShardColumn.Name.L
+	indexed := false
+	tbl, err := sessiontxn.GetTxnManager(se).GetTxnInfoSchema().TableByName(tableName.Schema, tableName.Name)
+	if err != nil {
+		return false, nil, err
+	}
+
+	if shardColumnName == "_tidb_rowid" && !tbl.Meta().PKIsHandle && !tbl.Meta().IsCommonHandle {
+		return true, nil, nil
+	}
+
+	var shardColumn *table.Column
+	for _, col := range tbl.Cols() {
+		if col.Name.L == shardColumnName {
```

Comment:
Not so sure. I didn't think of an anti-example for generated columns. As long as it can be effeciently indexed and queried, I think it works(?)
Anyway, we will need to test it.

Reply:
Do we need to consider table name of the column? such as a sql "split on test4.id limit 1000 dry run delete from test3  where id < 2000;". It was executed successfully, but it was not so reasonable.